### PR TITLE
many changes for PMacc :-)

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add `AlpakaHost`
- allocate host buffers on device `AlpakaHost`
- add missing event system hints
- ...

@BenjaminW3  Please check my changes, if anything is wrong please mark it and I change it.

### CODE REVIEW

File: src/libPMacc/examples/gameOfLife2D/src/main.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@
  * If not, see <http://www.gnu.org/licenses/>.
  */
 
+#define ALPAKA_DEBUG ALPAKA_DEBUG_FULL
```
Comment (BenjaminW3) [inline src/libPMacc/examples/gameOfLife2D/src/main.cpp]: This can be removed.

File: src/picongpu/main.cpp [hunk 2]
```diff
@@ -70,7 +73,7 @@ using AllocatorHostNew = mallocMC::Allocator<
     mallocMC::DistributionPolicies::Noop,
     mallocMC::OOMPolicies::ReturnNull,
     mallocMC::ReservePoolPolicies::NoOp,
-    mallocMC::AlignmentPolicies::Shrink<>
+    mallocMC::AlignmentPolicies::Noop
```
Comment (BenjaminW3) [inline src/picongpu/main.cpp:75]: Is this change necessary?

File: src/picongpu/main.cpp [hunk 3]
```diff
@@ -63,6 +64,8 @@ MALLOCMC_SET_ALLOCATOR_TYPE( ScatterAllocator );
 
 #else
 
+#define PMACC_ACC_CPU
```
Comment (BenjaminW3) [inline src/picongpu/main.cpp]: Please remove this, I have moved this to PMacc.

File: src/picongpu/main.cpp (hunk 2 above)
Comment (psychocoderHPC) [inline src/picongpu/main.cpp:75]: yes `Shrink<>` is not allowed on the host allocator

File: src/picongpu/main.cpp (hunk 3 above)
Comment (psychocoderHPC) [inline src/picongpu/main.cpp]: can you please link to the place were it is defined?

File: src/libPMacc/include/types.h [hunk 4]
```diff
@@ -44,13 +44,11 @@ namespace PMacc
     using AlpakaSize = size_t;
     using AlpakaIdxSize = int;
 
-#if !defined(__CUDACC__)
-    #define PMACC_ACC_CPU
-#endif
-
+    using AlpakaHost = alpaka::dev::DevCpu;
```
Comment (BenjaminW3) [inline src/libPMacc/include/types.h:51]: Could you rebase this commit and not remove the device detection? I had already corrected this in between.

File: src/libPMacc/include/nvidia/memory/MemoryInfo.hpp [hunk 5]
```diff
@@ -57,6 +57,7 @@ class MemoryInfo
                 alpaka::dev::getFreeMemBytes(
                     *device.get()));
 
+            freeInternal= (freeInternal * 3) / 4;
```
Comment (BenjaminW3) [inline src/libPMacc/include/nvidia/memory/MemoryInfo.hpp]: Is this still required?

File: src/picongpu/main.cpp (hunk 3 above)
Comment (BenjaminW3) [inline src/picongpu/main.cpp]: It is deleted in the comment below. I already had added the (hopefully) correct version.

**Comment by BenjaminW3** [comment]
Thank you very much! Many things I had overlooked! Especially the Host<->Device mix can explain many crashes ;-)

**Comment by psychocoderHPC** [comment]
updated

File: src/libPMacc/include/nvidia/memory/MemoryInfo.hpp [hunk 6]
```diff
@@ -89,6 +92,7 @@ class MemoryInfo
 
         /* alloc 90%, since allocating 100% is a bit risky on a SoC-like device */
         size_t allocSth = size_t( 0.9 * double(freeAtStart) );
+        std::cout<<"try to allocate "<<allocSth<<" "<<allocSth/1024/1024<<std::endl;
```
Comment (BenjaminW3) [inline src/libPMacc/include/nvidia/memory/MemoryInfo.hpp]: Is this debug message necessary?

File: src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp [hunk 7]
```diff
@@ -81,11 +82,12 @@ class ParticlesBox : protected DataBox<PitchedBox<SuperCell<FRAME>, DIM> >
      */
     DINLINE FRAME &getEmptyFrame() const
     {
+
         FrameType* tmp = NULL;
         const int maxTries = 13; //magic number is not performance critical
         for (int numTries = 0; numTries < maxTries; ++numTries)
         {
-            tmp = (FrameType*) mallocMC::malloc(sizeof (FrameType));
+            tmp = (FrameType*) malloc(sizeof(FrameType)); //mallocMC::malloc(sizeof (FrameType));
```
Comment (BenjaminW3) [inline src/libPMacc/include/particles/memory/boxes/ParticlesBox.hpp]: Is mallocMC currently infunctional?

**Comment by psychocoderHPC** [comment]
updated again^^

File: src/libPMacc/include/mpi/MPIReduce.hpp [hunk 8]
```diff
@@ -52,6 +52,7 @@ struct MPIReduce
     {
         if (isMPICommInitialized)
         {
+            /* exception inside the destructor are forbidden */
```
Comment (BenjaminW3) [inline src/libPMacc/include/mpi/MPIReduce.hpp:55]: They are not really forbidden, but should be assumed forbidden as a general rule because throwing exceptions in the destructor is even worse then goto ;-)

File: src/libPMacc/examples/gameOfLife2D/src/main.cpp [hunk 9]
```diff
@@ -19,6 +19,7 @@
  * If not, see <http://www.gnu.org/licenses/>.
  */
 
+
```
Comment (BenjaminW3) [inline src/libPMacc/examples/gameOfLife2D/src/main.cpp]: Empty line?

File: src/picongpu/include/particles/ParticlesInit.kernel [hunk 10]
```diff
@@ -189,6 +189,7 @@ ALPAKA_FN_ACC void operator()(
         if (linearThreadIdx == 0 && finished == 0)
         {
             frame = &(pb.getEmptyFrame());
+            std::cout<<"ptr "<<frame<<std::endl;
```
Comment (BenjaminW3) [inline src/picongpu/include/particles/ParticlesInit.kernel]: debug log...

File: src/libPMacc/include/mpi/MPIReduce.hpp (hunk 8 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/mpi/MPIReduce.hpp:55]: I added this commend because it was not clear why `NOEXEPT` was used and not the old version. We should leave this comment in for now that it is easier for me to review it later if it goes back to the mainline of PIConGPU

Comment (BenjaminW3) [inline src/libPMacc/include/mpi/MPIReduce.hpp:55]: My point did not get clear: I also meant that the comment does not have to be changed. It was only a clarification.

Comment (psychocoderHPC) [inline src/libPMacc/include/mpi/MPIReduce.hpp:55]: kk

**Comment by psychocoderHPC** [comment]
updated
